[FRONTEND] Implement AddressList component in the payment page

ozdenizdolu opened the pull request:
## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Description

AddressList component (similar to CreditCardList) has been implemented for the payment page. It looks as below screenshots:

# Screenshots

![image](https://user-images.githubusercontent.com/32651175/101986991-7463d400-3ca2-11eb-81ab-b7f6ce6fd0f2.png)
![image](https://user-images.githubusercontent.com/32651175/101987008-85ace080-3ca2-11eb-9810-f6eaae3970e9.png)
![image](https://user-images.githubusercontent.com/32651175/101987011-8fcedf00-3ca2-11eb-9d3a-a1fadb600ea8.png)
![image](https://user-images.githubusercontent.com/32651175/101987019-99f0dd80-3ca2-11eb-9504-c7f593cb1b3a.png)


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks
I have implemented this part of the code by refactoring the code for the CreditCardList. Therefore, there might be some potential problems in the code where I forgot to refactor correctly. I haven't detected any.

Closes #231

Commit:
Started working on addresslist component on checkout
- Changed the folder structure of components/checkout (added 2 folders
  cardlist & addresslist in checkout)
- Minor corrections on creditcardlist
- Started creating address cards

Commit:
disable payment button when shopping cart empty

Commit:
Merge remote-tracking branch 'origin/master' into F-checkout-addresslist

Commit:
minor styling changes to address card

Commit:
AddressList component has been completed

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressCard.js:
```diff
@@ -0,0 +1,88 @@
+            onAddressInfoChange();
+            notification.success({ message: 'You have successfully deleted your address!' })
+            
+        } catch (error) {
+            notification.warning({ message: 'There was an error with your request.' })
+        } finally {
```
print the error to the console too for debugging purposes

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressCard.js:
```diff
@@ -0,0 +1,88 @@
+                        <div className="address-card-name">{address.name + ' ' + address.surname}</div>
+                        <div className="address-card-address">{address.address}</div>
+                        <div className="address-card-zipcode">{address.zipCode}</div>
+                        <div className="address-card-region">{address.province + ' / ' + address.city + ' / ' + address.country}</div>
+                        <div className="address-card-phone">{address.phone.countryCode + address.phone.number}</div>
```
```js
[address.province, address.city, address.country].filter(Boolean).join('/')
```
use this instead to avoid showing falsy values

you can do this also for phone country code + phone number

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressList.js:
```diff
@@ -0,0 +1,113 @@
+            <div className="addresslist-list">
+                <Spin spinning={isLoading}>
+                    <List
+                        locale={{emptyText: "Add a new payment option!"}}
```
replace text for address

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressList.less:
```diff
@@ -0,0 +1,27 @@
+    margin-right: 15px;
+    padding: 15px;
+
+    //display: grid;
```
delete comment

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressModalInner.js:
```diff
@@ -0,0 +1,171 @@
+            layout="horizontal"
+            form={form}
+            name={"addressForm"+address?.id}
+            onFinish={() => {}}
+            onFinishFailed={() => {}}
```
no need to pass empty functions

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressModalInner.js:
```diff
@@ -0,0 +1,171 @@
+            onFinish={() => {}}
+            onFinishFailed={() => {}}
+            scrollToFirstError
+            initialValues={address ?? {...R.omit(['id'], address)}}>
```
initialValues should be given correctly from parent component because it has access to the mode.

MehdiSaffar replied:
also no need to bother with R.omit

MehdiSaffar replied:
also no need to spread R.omit(..)

MehdiSaffar commented on ecommerce/frontend/src/components/checkout/addresslist/AddressModalInner.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+
+    return <>
+        <Form
```
No need for `<>` fragment since the only child is `<Form>` element

MehdiSaffar committed:
fix login successful

MehdiSaffar reviewed (changes requested):
Very well done, I just thought of the following changes:

- remove alert when pressing proceed to payment
- instead of doing margin-{top,left,right} make parent component display: flex and use gap: 8px instead
- make the price summary slide down (with position:sticky) as I showed you in the meeting
- phone number input should be regular input (otherwise problem with big numbers) also add validation to it (at least check non-empty + only numbers)
- same thing for zip code

abdullahyldz commented on ecommerce/frontend/src/components/checkout/addresslist/AddressCard.js:
```diff
@@ -0,0 +1,76 @@
+import { Popconfirm, Spin, notification } from 'antd'
+import { sleep } from '../../../utils'
+
+export const AddressCard = ({ address, selected = false, onSelect = () => {}, onAddressInfoChange = () => {} }) => {
+    const [editVisible, setEditVisible] = useState(false)
```
making the `div `wrapped with` onSelect` is a super nice user experience, well done.

abdullahyldz replied:
Also, clicking `delete `or `edit `icons select the item. This is bad :)

ozdenizdolu replied:
I noticed that it selects on clicking edit&delete. I can see it's not the best behavior but I also think that it's somehow harmless. Can you see a harmful outcome out of this?

MehdiSaffar replied:
easy way is to call event.stopPropagation() on child onClick
https://stackoverflow.com/questions/37568550/react-prevent-event-trigger-on-parent-from-child

abdullahyldz commented on ecommerce/frontend/src/components/checkout/addresslist/AddressCard.js:
```diff
@@ -0,0 +1,76 @@
+                    <div className="address-card-region">
+                        {address.province + ' / ' + address.city + ' / ' + address.country}
+                    </div>
+                    <div className="address-card-phone">{address.phone.countryCode + address.phone.number}</div>
+                </div>
+            </Spin>
+        </div>
```
Can you also add at the bottom "red star next to input fields means they are 'necessary to fill'" for the sake of clearness?

ozdenizdolu replied:
added it on address edit form, did not add it on credit card form due to aesthetic reasons

abdullahyldz commented on ecommerce/frontend/src/components/checkout/addresslist/AddressModal.js:
```diff
@@ -0,0 +1,60 @@
+            visible={visible}
+            title={mode === 'add' ? 'Add a new address' : 'Edit your address'}
+            destroyOnClose
+            onOk={mode === 'add' ? onAdd : onEdit}
+            onCancel={onCancel}
+            cancelText="Go Back"
+            okText={mode === 'add' ? 'Add' : 'Edit'}>
+            <Spin spinning={isLoading}>
```
'Cancel' instead of 'Go Back' is better.

abdullahyldz reviewed (approved):
Very well designed work.

Commit:
Changed the folder structure(got rid of /checkout/)

Commit:
Merge remote-tracking branch 'origin/master' into F-checkout-addresslist

Commit:
Now, the payment button don't alert total price

Commit:
Address/card list have been put in a container

Commit:
Create the layout(containers etc) in checkoutPage

Commit:
Changes made according to Abdullah's comments
- Address edit modal now shows "* marked fields are required"
- Cancel button says "Cancel" on modals instead of "Go Back"

Commit:
Empty text for addressList is now correct

Commit:
Made the final changes for Mehdi's comments:
- Deleted the unnecessary <> fragment
- Got rid of R.omit in the AddressModalInner
- Deleted the unnecessary empty functions in form
- Deleted unnecessary comments here and there

MehdiSaffar commented on ecommerce/frontend/src/components/ShoppingCart/ShoppingCartPage.js:
```diff
@@ -57,8 +57,8 @@ export const ShoppingCartPage = ({ currency = 'TL' }) => {
                 <div className="shopping-proceed-button">
-                    <Button type="primary" onClick={onCheckout} block>
-                        Proceed to payment
+                    <Button type="primary" onClick={onCheckout} block disabled={shoppingCart?.length === 0}>
+                        <Link to="/checkout">Proceed to payment</Link>
                     </Button>
```
not sure if it makes sense to have both an onClick and a link? if you have complex logic then it is enough to have onClick which checks for some things and if okay uses history.push(url) to go to /checkout

MehdiSaffar commented on ecommerce/frontend/src/components/addresslist/AddressCard.js:
```diff
@@ -0,0 +1,76 @@
+                    <div className="address-card-zipcode">{address.zipCode}</div>
+                    <div className="address-card-region">
+                        {address.province + ' / ' + address.city + ' / ' + address.country}
+                    </div>
```
any decision on this btw?

MehdiSaffar commented on ecommerce/frontend/src/components/addresslist/AddressList.js:
```diff
@@ -0,0 +1,117 @@
+                    <List
+                        locale={{ emptyText: 'Add a new payment option!' }}
+                        grid={{ gutter: 0 }}
+                        dataSource={addressList}
+                        renderItem={address => (
+                            <List.Item>
+                                <AddressCard
+                                    address={address}
+                                    onAddressInfoChange={onAddressInfoChange}
+                                    selected={selectedAddress === address.id}
+                                    onSelect={onAddressSelect}
+                                />
+                            </List.Item>
+                        )}
+                    />
```
there needs to be itemKey or key prop to List.Item here.

MehdiSaffar commented on ecommerce/frontend/src/components/addresslist/AddressModalInner.js:
```diff
@@ -0,0 +1,191 @@
+                onFinish={() => {}}
+                onFinishFailed={() => {}}
+                scrollToFirstError
+                initialValues={address ?? { ...R.omit(['id'], address) }}>
```
shouldn't we leave id in there? (think of edit mode)

ozdenizdolu commented:
I have implemented all the requested changes. Only problem is: I could not make `position:sticky;`  work for payment summary part.

MehdiSaffar commented on ecommerce/frontend/src/components/addresslist/AddressModalInner.js:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    return (
+        <>
```
no need for wrapping Fragment

MehdiSaffar reviewed (approved):
Looks good, just some tiny details need to be done

Commit:
merge conflict solved

Commit:
Finishing touches(Considering Mehdi's comments)
